### PR TITLE
Semantic: Allow macro overload on named args

### LINKED ISSUE #5769: Macro is overwritten by different named arguments
I thought there was another issue about this but couldn't find it.

When I make 2 macros with the same name but different arguments, the second definition overwrite the first definition.
This is normal when the difference is for positional arguments, but it also does that for _named argument_, which I think is wrong:
```cr
macro foo(*, arg1)
end

# The first definition is overwritten :/
macro foo(*, arg2)
end

foo arg1: 3 # => Error: missing argument: arg2
foo arg2: 3 # => compiles
```
(first call: https://carc.in/#/r/3o4c, second call: https://carc.in/#/r/3o4d)

As the 2 definitions have different named arguments, I think the second definition should be independent from the first one instead of overwriting it.

### PR DESCRIPTION
**Pull request by bew**
I think this is my first PR in the compiler, deep into the semantic pass :tada: 

Fixes #5769 

I added a few comments here and there to help me understand what some parts does, I can remove them if needed. I also fixed a bad argument name (`a_def` => `a_macro`) in 2 macro-related methods.

I'll add more specs later if you think what I did is correct.
Thank you!

### CODE REVIEW

File: spec/compiler/semantic/macro_overload_spec.cr [hunk 1]
```diff
@@ -0,0 +1,29 @@
+require "../../spec_helper"
+
+describe "Semantic: macro overload" do
+  it "something between 2 macros (I need a name!!!)" do
+    assert_type(%(
+      macro foo(*, arg1)
+        {{ arg1 }}
```
Comment (RX14) [inline spec/compiler/semantic/macro_overload_spec.cr]: Perhaps use `1` and `1_i64` for these two so the spec can tell if the compiler uses the other overload.

File: src/compiler/crystal/semantic/restrictions.cr [hunk 2]
```diff
@@ -193,12 +195,38 @@ module Crystal
...
     def overrides?(other : Macro)
-      # For now we consider that a macro overrides another macro
-      # if it has the same number of arguments, splat index and
-      # named arguments.
-      args.size == other.args.size &&
-        splat_index == other.splat_index &&
-        !!double_splat == !!other.double_splat
+      # If they have different number of arguments, splat index or presence of
+      # double splat, no override.
+      unless args.size == other.args.size &&
```
Comment (RX14) [inline src/compiler/crystal/semantic/restrictions.cr]: I'd apply de morgan's here

```cr
if args.size != other.args.size ||
...
```

File: spec/compiler/semantic/macro_overload_spec.cr (hunk 1 above)
Comment (bew) [reply spec/compiler/semantic/macro_overload_spec.cr]: Rebased, and I used `1` and `"foo"`, as I find it easier to directly see the difference in type of these two literals!

File: src/compiler/crystal/semantic/restrictions.cr (hunk 2 above)
Comment (bew) [reply src/compiler/crystal/semantic/restrictions.cr]: done

**Comment by RX14** [comment]
needs a format

**Comment by bew** [comment]
Cool, thank you :smiley:

**Pull request merged**